src/node_file.h

```diff
@@ -148,6 +148,82 @@ class FSReqCallback : public FSReqBase {
   DISALLOW_COPY_AND_ASSIGN(FSReqCallback);
 };
 
+template <typename NativeT,
+          // SFINAE limit NativeT to arithmetic types
+          typename = std::enable_if<std::is_arithmetic<NativeT>::value>>
+constexpr NativeT ToNative(uv_timespec_t ts) {
+  // This template has exactly two specializations below.
+  static_assert(std::is_arithmetic<NativeT>::value == false, "Not implemented");
+  UNREACHABLE();
+}
+
+template <>
+constexpr double ToNative(uv_timespec_t ts) {
+  // We need to do a static_cast since the original FS values are ulong.
+  /* NOLINTNEXTLINE(runtime/int) */
+  const auto u_sec = static_cast<unsigned long>(ts.tv_sec);
+  const double full_sec = u_sec * 1000.0;
+  /* NOLINTNEXTLINE(runtime/int) */
+  const auto u_nsec = static_cast<unsigned long>(ts.tv_nsec);
+  const double full_nsec = u_nsec / 1000'000.0;
+  return full_sec + full_nsec;
+}
+template <>
+constexpr uint64_t ToNative(uv_timespec_t ts) {
+  // We need to do a static_cast since the original FS values are ulong.
+  /* NOLINTNEXTLINE(runtime/int) */
+  const auto u_sec = static_cast<unsigned long>(ts.tv_sec);
+  const auto full_sec = static_cast<uint64_t>(u_sec) * 1000UL;
+  /* NOLINTNEXTLINE(runtime/int) */
+  const auto u_nsec = static_cast<unsigned long>(ts.tv_nsec);
+  const auto full_nsec = static_cast<uint64_t>(u_nsec) / 1000'000UL;
+  return full_sec + full_nsec;
+}
+
+template <typename NativeT, typename V8T>
+constexpr void FillStatsArray(AliasedBuffer<NativeT, V8T>* fields,
+                              const uv_stat_t* s, size_t offset = 0) {
+  fields->SetValue(offset + 0, s->st_dev);
+  fields->SetValue(offset + 1, s->st_mode);
+  fields->SetValue(offset + 2, s->st_nlink);
+  fields->SetValue(offset + 3, s->st_uid);
+  fields->SetValue(offset + 4, s->st_gid);
+  fields->SetValue(offset + 5, s->st_rdev);
+#if defined(__POSIX__)
+  fields->SetValue(offset + 6, s->st_blksize);
+#else
+  fields->SetValue(offset + 6, 0);
+#endif
+  fields->SetValue(offset + 7, s->st_ino);
+  fields->SetValue(offset + 8, s->st_size);
+#if defined(__POSIX__)
+  fields->SetValue(offset + 9, s->st_blocks);
+#else
+  fields->SetValue(offset + 9, 0);
+#endif
+// Dates.
+  fields->SetValue(offset + 10, ToNative<NativeT>(s->st_atim));
+  fields->SetValue(offset + 11, ToNative<NativeT>(s->st_mtim));
+  fields->SetValue(offset + 12, ToNative<NativeT>(s->st_ctim));
+  fields->SetValue(offset + 13, ToNative<NativeT>(s->st_birthtim));
+}
+
+inline Local<Value> FillGlobalStatsArray(Environment* env,
+                                         bool use_bigint,
+                                         const uv_stat_t* s,
+                                         bool second = false) {
+  ptrdiff_t offset = second ? kFsStatsFieldsNumber : 0;
+  if (use_bigint) {
+    const auto arr = env->fs_stats_field_bigint_array();
```

Comment:
I understand the desire to not spell out the full type here, but could you make it `const auto*`? That way it’s obvious that no copy operations take place here, not knowing the type of `arr`

Reply:
I wanted to, but the compiler was being strage, I'll check again.

Reply:
```
error C2664: 'void node::fs::FillStatsArray<uint64_t,v8::BigUint64Array>(node::AliasedBuffer<uint64_t,v8::BigUint64Array> *,const uv_stat_t *,size_t)': cannot convert argument 1 from 'const node::AliasedBuffer<ui
nt64_t,v8::BigUint64Array> *' to 'node::AliasedBuffer<uint64_t,v8::BigUint64Array> *' [D:\code\node\node_lib.vcxproj]
```
It's a covariance thing.
I could define a using for the two variants...

Reply:
It’s a `const` thing – `auto*`, not `const auto*` is what I should have written.